File: scala/src/main/scala/com/madhukaraphatak/scala/simple/ADT.scala
package com.madhukaraphatak.scala.simple
sealed trait Optional[+T]

case object CNone extends Optional[Nothing]

case class CSome[T](value:T) extends Optional[T]

object ADT {

  def main(args:Array[String]) {
     val intOptionalNone:Optional[Int] = CNone
     val intOptionalSome:Optional[Int] = CSome(10)

     intOptionalSome match {
     case CSome(value) => println("value is {}", value)
     case CNone => println("none value")
   }
}
}

File: rust/simple/src/bin/adt.rs
enum Optional<T> {
 CNone,
 CSome(T)
}

fn main() {
  let int_optional_none:Optional<i32> = Optional::CNone;
  let int_optional_some:Optional<i32> = Optional::CSome(10);

  match int_optional_some {
    Optional::CSome(value) => println!("value is {}", value),
    Optional::CNone => println!("none value")
   }
}
